### PR TITLE
(MERGE PR #50 BEFORE TOUCHING THIS) Table Component for Courses plus tests

### LINKED ISSUE #45: Table Component for Courses plus tests

### PR DESCRIPTION
**Pull request by adenjo6**
[Storybook](https://ucsb-cs156-f24.github.io/proj-happycows-f24-11/prs/51/storybook/?path=/story/components-courses-coursestable--three-items-ordinary-user)

## Overview

In this PR, I finished Courses Tables. I added CoursesTable.js, CoursesTable.tests.js, coursesUtils.js, and CoursesTable.stories.js

## Screenshots (Optional)
Storybook:
Screenshot 2024-11-25 at 1 33 07 PM

Empty Table:
Screenshot 2024-11-25 at 1 30 31 PM

Three Items (USER):
Screenshot 2024-11-29 at 4 22 43 PM

Three Items (ADMIN):
Screenshot 2024-11-29 at 4 22 57 PM

Mutation Testing:

Screenshot 2024-11-25 at 1 35 23 PM

## Tests

- [x] Backend Unit tests (`mvn test`) pass
- [x] Backend Test coverage (`mvn test jacoco:report`) 100%
- [x] Backend Mutation tests (`mvn test pitest:mutationCoverage`) 100% 
- [x] Frontend Unit tests (`npm test`) pass
- [x] Frontend Test coverage (`npm run coverage`) 100%
- [x] Frontend Mutation tests (`npx stryker run`) 100% 
- [x] Frontend Linting (`npx eslint --fix src`) 

## Linked Issues

Closes #45

### CODE REVIEW

**Review by smmzhu — DISMISSED** [review]
lgtm

**Comment by adenjo6** [comment]
Note:
I clicked merge instead of rebase, so my previous PR #50 's 16 files show up here. there are only 5 files changed for this specific PR.  I will make sure to rebase whenever possible to avoid clutter.

File: frontend/src/main/components/Courses/CoursesTable.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+             accessor: 'term',
+         },
+     ];
+
+     if (hasRole(currentUser, "ROLE_ADMIN")) {
+         columns.push(ButtonColumn("Edit", "primary", editCallback, "CoursesTable"));
+         columns.push(ButtonColumn("Delete", "danger", deleteCallback, "CoursesTable"));
+     }
+     
+     columns.push(ButtonColumn("Join", "primary", joinCallback, "CoursesTable"));
```
Comment (pconrad) [inline frontend/src/main/components/Courses/CoursesTable.js:52]:
```suggestion

```

File: frontend/src/main/components/Courses/CoursesTable.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ import { useNavigate } from "react-router-dom";
+ import { hasRole } from "main/utils/currentUser";
+
+ export default function CoursesTable({ courses, currentUser }) {
+
+     const navigate = useNavigate();
+
+     const joinCallback = (cell) => {
+        navigate(`/courses/join/${cell.row.values.id}`);
+    };
```
Comment (pconrad) [inline frontend/src/main/components/Courses/CoursesTable.js:14]: What is this 'join' page?   I don't think we have one in this app.  If I'm mistaken, please explain.
```suggestion

```

File: frontend/src/tests/components/Courses/CoursesTable.test.js [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      const header = screen.getByTestId(`${testId}-cell-row-0-col-${field}`);
+      expect(header).toBeInTheDocument();
+    });
+
+    expect(screen.getByTestId(`${testId}-cell-row-0-col-id`)).toHaveTextContent("1");
+    expect(screen.getByTestId(`${testId}-cell-row-1-col-id`)).toHaveTextContent("2");
+
+    const joinButton = screen.queryByTestId(`${testId}-cell-row-0-col-Join-button`);
+    expect(joinButton).toBeInTheDocument();
+    expect(joinButton).toHaveClass("btn-primary");
```
Comment (pconrad) [inline frontend/src/tests/components/Courses/CoursesTable.test.js:51]:
```suggestion

```

File: frontend/src/tests/components/Courses/CoursesTable.test.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    expectedFields.forEach((field) => {
+      const header = screen.getByTestId(`${testId}-cell-row-0-col-${field}`);
+      expect(header).toBeInTheDocument();
+    });
+
+    expect(screen.getByTestId(`${testId}-cell-row-0-col-id`)).toHaveTextContent("1");
+    expect(screen.getByTestId(`${testId}-cell-row-1-col-id`)).toHaveTextContent("2");
+
+    const joinButton = screen.queryByTestId(`${testId}-cell-row-0-col-Join-button`);
+    expect(joinButton).toBeInTheDocument();
```
Comment (pconrad) [inline frontend/src/tests/components/Courses/CoursesTable.test.js:131]:
```suggestion
 
```

**Review by pconrad — CHANGES_REQUESTED** [review]
What is the Join button for that you can see in this [Storybook](https://ucsb-cs156-f24.github.io/proj-happycows-f24-11/prs/51/storybook/?path=/story/components-courses-coursestable--three-items-ordinary-user) and: Screenshot:

image

My guess is that this is left over from the Organic code you copied over—but here's the thing: there's no "join" functionality in HappyCows, unlike Organic—no user story that this button connects to, and no backend endpoint that would make sense for it to be connected to.

Always be thinking about what the end user is going to do with the code you are adding!

**Review by adenjo6 — COMMENTED** [review]
Accepted all changes. The join button should be removed as the Courses will be implemented in the commons in the future

**Pull request merged**